src/puppetlabs/puppetdb/command.clj

```diff
@@ -428,20 +414,27 @@
         (recur (dec n))
         result))))
 
-(defn process-command-and-respond! [cmd db response-pub-chan stats-atom]
+(defn process-command-and-respond! [{:keys [callback] :as cmd} db response-pub-chan stats-atom]
   (try
     (let [result (call-with-quick-retry 4
                   (fn []
                     (process-command! cmd db)))]
       (swap! stats-atom update :executed-commands inc)
+      (callback {:command cmd :result result})
       (async/>!! response-pub-chan
                  (make-cmd-processed-message cmd nil))
       result)
     (catch Exception ex
+      (callback {:command cmd :exception ex})
```

Comment:
I haven't figured out the full context yet, but just wanted to make sure we thought this was however bulletproof we need it to be, i.e. do we assume the callback doesn't throw (and is that made clear to the callback providers), or is it fine if it does?